### PR TITLE
Module API Work

### PR DESCRIPTION
**Pull request by rabbitfang**
See `docs/module-api.md` for some notes when you are reviewing.

### CODE REVIEW

File: modules/aws_base/input_schema.aws_base.json
Comment (RemyDeWolf) [inline modules/aws_base/input_schema.aws_base.json:2]: how can we merge this without having a reference to localhost?
do we even need this to be a URL?
The existing id are actually not valid URL
```
curl https://app.runx.dev/modules/aws-base
curl: (6) Could not resolve host: app.runx.dev
```

File: modules/aws_base/input_schema.aws_base.json
Comment (RemyDeWolf) [inline modules/aws_base/input_schema.aws_base.json:1]: to reduce maintenance while we have both implementation, would it be possible to reuse the existing json file? (`modules/aws_base/aws-base.json`)

File: opta/commands/apply.py
Comment (RemyDeWolf) [inline opta/commands/apply.py:137]: can we display a warning in the like - _not for production use_ or _beta feature_ to prevent unintentional use until it's ready?
(probably inside the new process.apply function itself)

File: opta/utils/features.py
Comment (RemyDeWolf) [inline opta/utils/features.py:18]: you can also add `y` for lazy folks

File: opta/core/cloud_provider.py
Comment (RemyDeWolf) [inline opta/core/cloud_provider.py:43]: note: I had to change my env file for it to work:
```
# opta.yaml
name: staging # name of the environment
org_name: runx
providers:
  aws:
    region: us-east-1
    #before
    # account_id: 248233625043
    account_id: [248233625043]
modules:
  - type: aws-base
```

File: opta/core/cloud_provider.py
Comment (RemyDeWolf) [inline opta/core/cloud_provider.py:20]: what do you think of breaking up the cloud implementation in it's own class from the beginning?
mostly to avoid having large files and a good separation between cloud specific impl / cloud agnostic

File: opta/core/cloud_provider.py
Comment (RemyDeWolf) [inline opta/core/cloud_provider.py:9]: this is an opportunity to document a bit, do you mind adding a one liner for each cloud function to implement?

File: opta/core/terraform2.py
Comment (RemyDeWolf) [inline opta/core/terraform2.py:13]: should we start putting the terraform code in a separate package and break into smaller classes?
It could be easier to test
ex: 
```
core.terraform.Terraform
core.terraform.TerraformFile
```

File: opta/core/cloud_provider.py
Comment (RemyDeWolf) [inline opta/core/cloud_provider.py:16]: is there a use case to call this method more than once?
we could rename this to `configure_providers` to reflect this

File: opta/core/cloud_provider.py
Comment (RemyDeWolf) [inline opta/core/cloud_provider.py:23]: this todo should be addressed before merge to achieve `aws-base` support

File: opta/core/cloud_provider.py
Comment (RemyDeWolf) [inline opta/core/cloud_provider.py:21]: I could not find where was this function used

File: opta/core/terraform2.py
Comment (RemyDeWolf) [inline opta/core/terraform2.py:45]: did you mean? Cloud _not_ copy ...

**Review by RemyDeWolf — COMMENTED** [review]
I will do more round of reviews, but here is some feedback for now.

Also I did some testing and while the 1st run worked, the 2nd run tried to recreate the full stack again. Adding test case below:

```
# opta.yaml
name: staging # name of the environment
org_name: runx
providers:
  aws:
    region: us-east-1
    account_id: [248233625043]
modules:
  - type: aws-base
```

First run, looks good. Except it's missing the terraform output which we had before
```
opta apply
...
Plan: 50 to add, 0 to change, 0 to destroy.
...
Apply complete! Resources: 50 added, 0 changed, 0 destroyed.
# missing terraform output here
```

Second run, I would expect no change, but it tried to recreate everything, issue with getting the state?
```
opta apply
...
Plan: 50 to add, 0 to change, 0 to destroy.
...
╷
│ Warning: Applied changes may be incomplete
│ 
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run the following command to verify that
│ no other changes are pending:
│     terraform plan
│ 	
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to use it as part
│ of an error message.
╵
╷
│ Error: error creating ElastiCache Subnet Group (opta-staging): CacheSubnetGroupAlreadyExists: Cache subnet group opta-staging already exists.
│ 	status code: 400, request id: 7ceba5cd-6e6d-417e-a9cd-502ddfb0f259
│ 
│   with module.aws-base.aws_elasticache_subnet_group.main,
│   on .terraform/modules/aws-base/cache_subnet.tf line 1, in resource "aws_elasticache_subnet_group" "main":
│    1: resource "aws_elasticache_subnet_group" "main" {
│ 
╵
╷
│ Error: Error creating DB Subnet Group: DBSubnetGroupAlreadyExists: The DB subnet group 'opta-staging' already exists.
│ 	status code: 400, request id: 35132597-d655-4333-8f75-6994cea7cc1d
│ 
│   with module.aws-base.aws_db_subnet_group.main,
│   on .terraform/modules/aws-base/db_subnet.tf line 1, in resource "aws_db_subnet_group" "main":
│    1: resource "aws_db_subnet_group" "main" {
│ 
╵
╷
│ Error: error creating DocDB Subnet Group: DBSubnetGroupAlreadyExists: The DB subnet group 'opta-staging-docdb' already exists.
│ 	status code: 400, request id: 26f28a8d-8040-4297-b8dc-5628e35bc8d9
│ 
│   with module.aws-base.aws_docdb_subnet_group.main,
│   on .terraform/modules/aws-base/documentdb_subnet.tf line 1, in resource "aws_docdb_subnet_group" "main":
│    1: resource "aws_docdb_subnet_group" "main" {
│ 
╵
╷
│ Error: error creating KMS Alias (alias/opta-staging): AlreadyExistsException: An alias with the name arn:aws:kms:us-east-1:248233625043:alias/opta-staging already exists
│ 
│   with module.aws-base.aws_kms_alias.alias,
│   on .terraform/modules/aws-base/kms.tf line 122, in resource "aws_kms_alias" "alias":
│  122: resource "aws_kms_alias" "alias" {
│ 
╵
╷
│ Error: Error creating EIP: AddressLimitExceeded: The maximum number of addresses has been reached.
│ 	status code: 400, request id: 043fc55a-090b-40ce-a91a-6c5152d1d8da
│ 
│   with module.aws-base.aws_eip.nat_eips[0],
│   on .terraform/modules/aws-base/private_subnet.tf line 31, in resource "aws_eip" "nat_eips":
│   31: resource "aws_eip" "nat_eips" {
│ 
╵
╷
│ Error: error creating IAM Role (opta-foo-vpc-flow): EntityAlreadyExists: Role with name opta-foo-vpc-flow already exists.
│ 	status code: 409, request id: 379b210d-dd69-4253-80e2-3901d5bfa3a2
│ 
│   with module.aws-base.aws_iam_role.vpc_flow_log,
│   on .terraform/modules/aws-base/vpc.tf line 33, in resource "aws_iam_role" "vpc_flow_log":
│   33: resource "aws_iam_role" "vpc_flow_log" {

```

File: opta/linker.py
Comment (RemyDeWolf) [inline opta/linker.py:30]: can this be covered by unit test?
sounds like a critical component of the new module api

File: opta/linker.py
Comment (RemyDeWolf) [inline opta/linker.py:58]: unit tests checking the execution order for different scenarios would help understand this code

File: opta/linker.py
Comment (RemyDeWolf) [inline opta/linker.py:44]: It's a bit overwhelming the amount of TODO in this PR :) 

Do you think we can introduce some prioritization?

You can define some high level milestones in docs/module-api.md
ex:
- milestone 1: get aws-base working creation
- milestone 2: get aws-base deletion
- milestone 3: get aws-base interpolation
- ... X: deprecate old module api

And then we can revisit the TODO to include the milestone where it will be needed.
- TODO M3:  Make sure that there are no interpolations in the link params

That will give some context and allow the rest of the team to start implementing these TODO
What do you think?

File: docs/module-api.md
Comment (RemyDeWolf) [inline docs/module-api.md:8]: What do you think should be the milestones to reach deprecation of the module API?
If you can add it in this document it will give some context about the next steps.
Also see my comment about the TODO to link them to these milestones.

File: opta/utils/visit.py
Comment (RemyDeWolf) [inline opta/utils/visit.py:1]: another critical class that you can add a few unit tests - to help with having the team understanding and enable us to make some changes

**Review by RemyDeWolf — COMMENTED** [review]
New batch of comments. 
Theme: add milestones, review TODO and more unit tests :)

File: modules/aws_base/module.yaml
Comment (juandiegopalomino) [inline modules/aws_base/module.yaml:5]: The previous id was of the form
`  "$id": "http://localhost:7800/modules/aws_base/input_schema.aws-base.json",`
why aren't we using localhost here?

File: opta/loader.py
Comment (juandiegopalomino) [inline opta/loader.py:29]: if we already have a jsonschema, then this should be straight forward, yes?

**Review by juandiegopalomino — COMMENTED** [review]
I know this is the pot calling the kettle doc, but could we have some docstrings?

File: opta/commands/apply.py
Comment (rabbitfang) [reply opta/commands/apply.py:137]: Added a warning log in the `process.apply` function.

File: opta/core/cloud_provider.py
Comment (rabbitfang) [reply opta/core/cloud_provider.py:1]: I think the distinction should be what in "internal" to opta vs what module processors are able to interact with. I agree that the current "core" separation is not well defined and is something that we should discuss (in a larger conversation on how the opta repo is organized).

File: opta/core/cloud_provider.py
Comment (rabbitfang) [reply opta/core/cloud_provider.py:21]: It's currently unused, but will be used shortly once remote state is implemented for module API.

File: opta/core/cloud_provider.py
Comment (rabbitfang) [reply opta/core/cloud_provider.py:20]: Once I dive deeper into the cloud specific code (such as remote state storage), I will refactor the python modules/classes and organize them better.

File: opta/core/cloud_provider.py
Comment (rabbitfang) [reply opta/core/cloud_provider.py:43]: This should be corrected now

File: opta/core/terraform2.py
Comment (rabbitfang) [reply opta/core/terraform2.py:13]: Separated these into their own files

File: opta/core/terraform2.py
Comment (rabbitfang) [reply opta/core/terraform2.py:45]: Fixed

File: opta/loader.py
Comment (rabbitfang) [reply opta/loader.py:29]: Yes, it should be pretty straight forward, but we do need to bypass some of the validation depending on where we are in the apply process.

File: opta/utils/features.py
Comment (rabbitfang) [reply opta/utils/features.py:18]: Added

File: opta/utils/visit.py
Comment (rabbitfang) [reply opta/utils/visit.py:1]: Adding in a separate PR to avoid blocking this PR.

**Review by rabbitfang — COMMENTED** [review]
There is currently a bug where local state doesn't persist between applys. Working on fixing that

File: modules/aws_base/input_schema.aws_base.json
Comment (RemyDeWolf) [reply modules/aws_base/input_schema.aws_base.json]: _We could add support to allow referencing the old file to the new YAML format._
that sounds like a good idea, and this would address the concern about maintaining two almost identical json files

**Review by juandiegopalomino — APPROVED** [review]

**Review by RemyDeWolf — COMMENTED** [review]
can you add a header in all the python files that are only used by module api?
ex:
```
# v2
```
```
# new-module-api
```
this will help with the transition period for us who are not familiar with the code base (easier that looking in github history) .
eventually we would remove them.

**Review by bigbitbus — APPROVED** [review]
Do you think we can add a doc string on some of these files to give a hint to the user about what the code in that file does at a high level; I know reading the code will be best, but it would be a big time commit for everyone on the team to go over the code and get a full understanding ...or you could consider a code walk-through like the Loom video @RemyDeWolf did for the terraform generator.

**Review by RemyDeWolf — COMMENTED** [review]
can we have a unit test to load a layer using a dummy fixture with only aws-base such as https://github.com/run-x/opta/blob/main/tests/fixtures/dummy_data/aws_env_getting_started.yaml
and:
- load the layer
- test terraform gen

**Comment by rabbitfang** [comment]
Going to split up this PR into smaller pieces

**Pull request closed without merging**
